api/server/user/data.go

```diff
@@ -176,63 +176,172 @@ func (s *Service) CidInfo(ctx context.Context, req *userPb.CidInfoRequest) (*use
+		if jobTimeA > jobTimeB {
+			return true
+		}
+		if jobTimeA < jobTimeB {
+			return false
+		}
+		return currentStorageTimeA > currentStorageTimeB
 	})
-	return &userPb.CidInfoResponse{CidInfos: res}, nil
+
+	summaries := make([]*userPb.CidSummary, len(sources))
+	for i, source := range sources {
+		summaries[i] = &userPb.CidSummary{
+			Cid:    source.cid,
+			Stored: source.currentStorage != nil,
+		}
+		if source.executingJob != nil {
+			summaries[i].ExecutingJob = source.executingJob.ID.String()
+		}
+		if len(source.queuedJobs) > 0 {
+			queuedIds := make([]string, len(source.queuedJobs))
+			for j, job := range source.queuedJobs {
+				queuedIds[j] = job.ID.String()
+			}
+			summaries[i].QueuedJobs = queuedIds
+		}
+	}
+
+	return &userPb.CidSummaryResponse{CidSummary: summaries}, nil
+}
+
+// CidInfo returns information about cids managed by the FFS instance.
+func (s *Service) CidInfo(ctx context.Context, req *userPb.CidInfoRequest) (*userPb.CidInfoResponse, error) {
+	i, err := s.getInstanceByToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cid, err := fromProtoCid(req.Cid)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "parsing cid: %v", err)
+	}
+
+	storageConfigs, err := i.GetStorageConfigs(cid)
+	if err != nil {
+		code := codes.Internal
+		if err == api.ErrNotFound {
+			code = codes.NotFound
+		}
+		return nil, status.Errorf(code, "getting storage configs: %v", err)
+	}
+	config, ok := storageConfigs[cid]
+	if !ok {
+		log.Warnf("didn't find storage config for cid %s", cid.String())
+		return nil, status.Errorf(codes.Internal, "didn't find storage config for cid %s", cid.String())
+	}
+
+	rpcConfig := ToRPCStorageConfig(config)
+	cidInfo := &userPb.CidInfo{
+		Cid:                       cid.String(),
+		LatestPushedStorageConfig: rpcConfig,
+	}
+	info, err := i.Show(cid)
+	if err != nil && err != api.ErrNotFound {
+		return nil, status.Errorf(codes.Internal, "getting storage info: %v", err)
+	} else if err == nil {
+		cidInfo.CurrentStorageInfo = toRPCStorageInfo(info)
+	}
+	queuedJobs := i.QueuedStorageJobs(cid)
+	rpcQueudJobs := make([]*userPb.StorageJob, len(queuedJobs))
+	for i, job := range queuedJobs {
+		rpcJob, err := toRPCJob(job)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "converting job to rpc job: %v", err)
+		}
+		rpcQueudJobs[i] = rpcJob
+	}
+	cidInfo.QueuedStorageJobs = rpcQueudJobs
+	executingJobs := i.ExecutingStorageJobs(cid)
+	if len(executingJobs) == 1 {
+		// There is exactly one job in the slice because we specified a cid
+		// and there can be only one executing job per cid at a time.
+		rpcJob, err := toRPCJob(executingJobs[0])
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "converting job to rpc job: %v", err)
+		}
+		cidInfo.ExecutingStorageJob = rpcJob
+	} else if len(executingJobs) > 1 {
+		log.Warnf("received %d executing jobs when there should be 1", len(executingJobs))
```

Comment:
Is safe not to `break` or return err in this case?

Reply:
It is safer to return an error. Done.